File: kirun-js/src/engine/model/FunctionOutput.ts
import { KIRuntimeException } from '../exception/KIRuntimeException';
import { isNullValue } from '../util/NullCheck';
import { EventResult } from './EventResult';
import { FunctionOutputGenerator } from './FunctionOutputGenerator';

export class FunctionOutput {
    private fo: EventResult[];

    private index: number = 0;
    private generator?: FunctionOutputGenerator;

    public constructor(arg: EventResult[] | FunctionOutputGenerator) {
        if (isNullValue(arg)) throw new KIRuntimeException('Function output is generating null');

        if (Array.isArray(arg) && arg.length && arg[0] instanceof EventResult) {
            this.fo = arg as EventResult[];
        } else {
            this.fo = [];
            if (!Array.isArray(arg)) this.generator = arg as FunctionOutputGenerator;
        }
    }

    public next(): EventResult | undefined {
        if (!this.generator) {
            if (this.index < this.fo.length) return this.fo[this.index++];
            return undefined;
        }

        const er: EventResult | undefined = this.generator.next();
        if (er) this.fo.push(er);
        return er;
    }

    public allResults(): EventResult[] {
        return this.fo;
    }
}

File: kirun-py/src/kirun_py/model/function_output.py
from __future__ import annotations
from typing import List, Optional, Union

from kirun_py.exception.ki_runtime_exception import KIRuntimeException
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output_generator import FunctionOutputGenerator


class FunctionOutput:
    def __init__(self, arg: Union[List[EventResult], FunctionOutputGenerator]):
        if arg is None:
            raise KIRuntimeException('Function output is generating null')

        self._index: int = 0
        self._generator: Optional[FunctionOutputGenerator] = None

        if isinstance(arg, list):
            self._fo: List[EventResult] = arg
        else:
            self._fo = []
            self._generator = arg

    def next(self) -> Optional[EventResult]:
        if self._generator is None:
            if self._index < len(self._fo):
                result = self._fo[self._index]
                self._index += 1
                return result
            return None

        er = self._generator.next()
        if er is not None:
            self._fo.append(er)
        return er

    def all_results(self) -> List[EventResult]:
        return self._fo
